src/protocol/messages/sasl_auth.rs

```diff
@@ -0,0 +1,206 @@
+    fn read_versioned(reader: &mut R, version: ApiVersion) -> Result<Self, ReadVersionedError> {
+        let v = version.0 .0;
+        assert!(v <= 1);
+        Ok(Self {
+            mechanism: String_::read(reader)?,
+        })
+    }
+}
+
+impl<W> WriteVersionedType<W> for SaslHandshakeRequest
+where
+    W: Write,
+{
+    fn write_versioned(
+        &self,
+        writer: &mut W,
+        version: ApiVersion,
+    ) -> Result<(), WriteVersionedError> {
+        let v = version.0 .0;
+        assert!(v == 1);
+        self.mechanism.write(writer)?;
+        Ok(())
+    }
+}
+
+impl RequestBody for SaslHandshakeRequest {
+    type ResponseBody = SaslHandshakeResponse;
+    const API_KEY: ApiKey = ApiKey::SaslHandshake;
+    const API_VERSION_RANGE: ApiVersionRange =
+        ApiVersionRange::new(ApiVersion(Int16(1)), ApiVersion(Int16(1)));
+    const FIRST_TAGGED_FIELD_IN_REQUEST_VERSION: ApiVersion = ApiVersion(Int16(3));
+}
+
+#[derive(Debug)]
+pub struct SaslHandshakeResponse {
+    pub error_code: Option<ApiError>,
+    pub mechanisms: Array<String_>,
+}
+
+impl<R> ReadVersionedType<R> for SaslHandshakeResponse
+where
+    R: Read,
+{
+    fn read_versioned(reader: &mut R, version: ApiVersion) -> Result<Self, ReadVersionedError> {
+        let v = version.0 .0;
+        assert!(v == 1);
+        Ok(Self {
+            error_code: ApiError::new(Int16::read(reader)?.0),
+            mechanisms: Array::read(reader)?,
+        })
+    }
+}
+
+impl<W> WriteVersionedType<W> for SaslHandshakeResponse
+where
+    W: Write,
+{
+    fn write_versioned(
+        &self,
+        _writer: &mut W,
+        _version: ApiVersion,
+    ) -> Result<(), WriteVersionedError> {
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct SaslAuthenticateRequest {
+    pub auth_bytes: CompactBytes,
+    pub tagged_fields: Option<TaggedFields>,
+}
+
+impl SaslAuthenticateRequest {
+    pub fn new(sasl_config: &SaslConfig) -> Self {
+        let mut auth: Vec<u8> = Vec::new();
+        auth.push(0);
+        auth.extend(sasl_config.username.bytes());
+        auth.push(0);
+        auth.extend(sasl_config.password.bytes());
+        return Self {
+            auth_bytes: CompactBytes(auth),
+            tagged_fields: Some(TaggedFields::default()),
+        };
+    }
+}
+
+impl<R> ReadVersionedType<R> for SaslAuthenticateRequest
+where
+    R: Read,
+{
+    fn read_versioned(reader: &mut R, version: ApiVersion) -> Result<Self, ReadVersionedError> {
+        let v = version.0 .0;
+        assert!(v == 2);
+        Ok(Self {
+            auth_bytes: CompactBytes::read(reader)?,
+            tagged_fields: (v >= 2).then(|| TaggedFields::read(reader)).transpose()?,
+        })
+    }
+}
+
+impl<W> WriteVersionedType<W> for SaslAuthenticateRequest
+where
+    W: Write,
+{
+    fn write_versioned(
+        &self,
+        writer: &mut W,
+        version: ApiVersion,
+    ) -> Result<(), WriteVersionedError> {
+        let v = version.0 .0;
+        assert!(v == 2);
```

Comment:
Why is version 1 not supported?

Edit: reading <https://kafka.apache.org/protocol#sasl_handshake> again I think we can also support version 0 for the `SaslAuthenticate` message.

Reply:
The sentence you're quoting is about the `SaslHandshakeRequest` (for which I agree that we should only support version 1), not about the `SaslAuthenticateRequest`. The versions for these two request-response types are different. The version selection will be done by the `Messenger`.

Reply:
I realize that your comment was regarding `SaslAuthenticate`. The issue I ran in to here was that the type of the fields got modified through the different versions. For e.g. `auth_bytes` was `BYTES` in version 0 but become `COMPACT_BYTES` from version 1 onwards. Is there an example for this is handled anywhere else in the library?

Reply:
I see. TBH we don't have a beautiful solution of it. I would go for `Bytes` as a type within the struct and then do some type conversion during read/write. See this example:

https://github.com/influxdata/rskafka/blob/ad8de3ccd7183f0d380ef2bd65a6fc1ea74fffb9/src/protocol/messages/delete_records.rs#L63-L66

Here the `name` is an ordinary string, but is compact after a certain version. This is then handled here:

https://github.com/influxdata/rskafka/blob/ad8de3ccd7183f0d380ef2bd65a6fc1ea74fffb9/src/protocol/messages/delete_records.rs#L89-L93

I think you might want to introduce `CompactBytesRef` for that, so you don't have to copy the payload during type conversion. This is similar how it is done for `CompactString`:

https://github.com/influxdata/rskafka/blob/ad8de3ccd7183f0d380ef2bd65a6fc1ea74fffb9/src/protocol/primitives.rs#L362-L365

https://github.com/influxdata/rskafka/blob/ad8de3ccd7183f0d380ef2bd65a6fc1ea74fffb9/src/protocol/primitives.rs#L392-L399

(note how the "write" operation is only implemented once, because `CompactString` can use `CompactStringRef`)

https://github.com/influxdata/rskafka/blob/ad8de3ccd7183f0d380ef2bd65a6fc1ea74fffb9/src/protocol/primitives.rs#L401-L417

Reply:
Neat! I will follow a similar pattern.